### ISSUE TITLE
~51 CUDA-only tests have no marker, no documented way to run them and no results record; both known runs on real cards found defects (#776, and 9 failures from three defects on H100)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#833]
## The defect

Every CI job runs on `ubuntu-latest`, `windows-latest`, `macos-latest` or `macos-14` (`.github/workflows/*.yml`). None of them has an NVIDIA GPU, so every CUDA-dependent test skips in CI. That part is expected. What is missing is everything around it: there is no way to select these tests, no instruction to run them, and nowhere to record the result. In practice they run only when someone happens to put the suite on a new card, and both times that has happened it found something.

## Evidence

**No marker.** `pyproject.toml:261-265` registers `smoke`, `unit` and `integration`, and nothing for GPU.

**Ten private copies of one probe.**

- `_cuda()` in `test_issue327_logits_estimate.py:23`, `test_issue349_measured_fit.py:435`, `test_issue623_stream_pin_allocation.py:67`, `test_v07202.py:32`, `test_v07203.py:21`

- `_cuda_available()` in `test_issue385_stream_dtype.py:382`, `test_v07200.py:1204`, `test_v07204.py:27`, `test_v07300.py:203`

- `_has_cuda()` in `test_v0403_part_a.py:181`

Five of those files then define the identical `requires_cuda = pytest.mark.skipif(not _cuda(), reason="needs a CUDA device")`.

**No documentation.** `README.md:403` says `pytest tests/ -v   # unit tests (fast, no GPU)`. Nothing mentions a subset that only means something on a GPU, or how to run it.

**No record.** The only published run of the subset on a real card is `benchmarks/gate-h100-validation.md` (STEP 1, around lines 470-503):

```

$ /root/venv/bin/python -m pytest tests/test_v07200.py tests/test_v07202.py \

      tests/test_v07203.py tests/test_v07204.py -v --no-cov

9 failed, 419 passed, 65 warnings in 36.12s

```

That record traces the failures to three distinct defects, none of them visible to CI. The second known run was on an RTX 5070 Laptop (sm_120, driver 616.92), and it produced #776 (NF4 not bit-exact on sm_120).

## Reproduction

A collection-time plugin evaluated every `skipif` mark twice on the RTX 5070 Laptop box (torch 2.14.0+cu130): once normally, and once with `CUDA_VISIBLE_DEVICES=-1`. The diff of the two reports is **44 tests in 9 files**:

```

>    23  'needs a CUDA device'  files=5

>     8  'the #328 meta failure is CUDA-only'  files=1

>     6  'requires CUDA (layer streaming is a GPU feature)'  files=2

>     6  'peak VRAM needs CUDA'  files=1

>     1  'the fused-kernel window is a CUDA dispatch'  files=1

```

About 7 more skip imperatively in the test body when CUDA is absent (`test_v0403_part_a.py:134-171` x5, `test_v07120.py:667`, `test_v07202.py:424`); that count comes from a static read, not a run. Two go the other way and skip **when CUDA is present**, so CI runs them and a GPU box does not (`test_issue349_measured_fit.py:137`, `test_v07120.py:655`).

## Why it matters

These tests guard the claims the project publishes numbers about: layer-streaming bit-exactness, VRAM prediction, NF4. Both recorded runs on a new card found defects, and both happened by accident. A contributor with a GPU who wants to help has no command to run and nowhere to report the result.

## Proposed fix

1. Register `gpu: needs a CUDA device (skipped without one; run with pytest -m gpu)` in `pyproject.toml`. Add one `requires_cuda` in `tests/conftest.py` that applies both the marker and the skip, and delete the ten copies.

2. Document in `CONTRIBUTING.md`, plus one line under the README's test commands: `pytest -m gpu --no-cov -v`, and what a report should include (card, compute capability, driver, torch build, pass/fail list).

3. Add `benchmarks/gpu-test-runs.md` with one row per card and date: pass/fail/skip counts and links to issues filed. Seed it with the H100 run and the RTX 5070 Laptop / #776 run.

4. Optional: a `workflow_dispatch` job for a self-hosted GPU runner if one becomes available.

## Acceptance criteria

- [ ] `pytest -m gpu --collect-only -q` selects every test that skips under `CUDA_VISIBLE_DEVICES=-1`. A test compares the two sets; it fails on current code because there is no `gpu` marker.

- [ ] One CUDA probe in `conftest.py`, and a ratchet that fails if a test module defines its own.

- [ ] `CONTRIBUTING.md` documents the command and the report format.

- [ ] `benchmarks/gpu-test-runs.md` exists and holds at least the two known runs.

---

**Related, filed the same day:**

- #834 — test_v0530 test_apply_live_gated (FP8 and NVFP4) depend on the real card and on torchao being absent: with ...

**Comment by Srinivasan8888** [comment]
I'd like to take this. **I have no NVIDIA card** (Apple Silicon only), so I can't do a real `pytest -m gpu` run on a card for the record file. I'll seed `benchmarks/gpu-test-runs.md` only from the two runs that are already published, and I'll say that in the PR.

## Re-measured on `main` (`5ee1278`): it has grown since filing

Static read of `tests/`:

- **15 private CUDA probes**, up from the 10 listed: `_cuda()` ×9, `_cuda_available()` ×5, `_has_cuda()` ×1. The new ones are in `test_issue1005_peft021_review_round.py:43`, `test_issue901_failed_page_lock_recovery.py:41`, `test_issue901_pinned_arenas.py:34`, `test_issue968_bnb_nf4_path_divergence.py:91` and `test_issue974_direct_range_reads.py:52`, and some existing line numbers have moved (`test_issue623_stream_pin_allocation.py:94`, `test_v07200.py:1227`, `test_v07300.py:217`).
- **9 identical `requires_cuda = pytest.mark.skipif(not _cuda(), reason="needs a CUDA device")`** definitions, up from 5.
- **Imperative in-body skips on missing CUDA:** still `test_v0403_part_a.py:134-171` ×5, `test_v07120.py:667` and `test_v07202.py:424`, plus `test_stream_multi_gpu_guard.py:118` (fewer than 2 devices).
- **The inverse pair is unchanged:** `test_issue349_measured_fit.py:137` and `test_v07120.py:655` skip *with* CUDA. They'll stay unmarked, since `-m gpu` shouldn't select a test that skips on a GPU.

I also tried the dynamic version of your measurement (collect twice, with `torch.cuda.is_available` patched True vs False). **On Apple Silicon it's noisy:** 3 collection errors, and MPS/trl skips flip for unrelated reasons. So it isn't evidence, and I'm not quoting its count. The test for criterion 1 will be structural instead (below), which also works on a CPU-only CI runner.

## Shape

1. **The marker:** register `gpu: needs a CUDA device (skipped without one; run with pytest -m gpu --no-cov)` in `pyproject.toml`. `tests/conftest.py` gets one cached probe, `cuda_available()`, and a `pytest_runtest_setup` hook that skips `gpu`-marked tests without CUDA. The mechanism is the same one @dchaudhari7177's #1039 uses for `requires_symlink`, so the two compose: both add a line to the marker list and a clause to the same hook. Whichever lands second takes the trivial conflict, and I'll take it if that's this one.
2. **Delete the copies:** remove the 15 probes and 9 `requires_cuda` definitions. Each use becomes `@pytest.mark.gpu`. Reasons more specific than "needs a CUDA device" (`peak VRAM needs CUDA`, `the #328 meta failure is CUDA-only`, …) move into the marker's docstring or a comment, so no information is lost.
3. **Imperative skips:** convert the ones that need CUDA for the whole test to the marker. The ones that need CUDA *plus* something else (`v07120:667` needs bitsandbytes too) keep their extra condition, but the CUDA half goes through the shared probe.
4. **Ratchet test:** an AST scan fails if any module under `tests/` other than `conftest.py` calls `torch.cuda.is_available()` inside a `skipif` condition, a `pytest.skip` guard or a module-level probe function. Two controls come with it:
   - a reject-everything control, so the scan can't pass vacuously;
   - an allowlist for tests that *patch* `is_available` (mocks, not probes), each entry with a reason.

   A second test drives the hook with the probe patched both ways: `gpu` skipped without CUDA, run with it.
5. **Docs:** `CONTRIBUTING.md` gets the command and the report format (card, compute capability, driver, torch build, pass/fail/skip, issues filed), plus one README line under the test commands. `benchmarks/gpu-test-runs.md` is seeded with the H100 run from `gate-h100-validation.md` and the RTX 5070 Laptop run behind #776, with figures copied from those records and not re-derived.

**Other overlaps I checked:**

- **#1002 (@Areen-09):** adds a GPU-free layer under `test_issue901_failed_page_lock_recovery.py`'s `@requires_cuda` tests. If it lands first, I'll rebase and convert whatever `requires_cuda` it keeps. If this lands first, those tests will be spelled `@pytest.mark.gpu`.
- **#981 and #1034:** touch `test_v07120.py`/`test_v07202.py`, but not the probe lines.
- **#1019:** touches `test_v07200.py`, where the probe sits at `:1227`.

**Comment by Srinivasan8888** [comment]
Opened as #1050 (`Refs #833`). Three of the four acceptance items are covered. The fourth, a real `pytest -m gpu` run in `benchmarks/gpu-test-runs.md`, needs a CUDA card, and I built and verified this on an Apple M1. The PR body has:

- **The skip comparison:** every test that skipped for CUDA on `main` is selected by `-m gpu`, and nothing new skips.
- **Mutation testing:** 13 of 13 killed.
- **Removed test lines:** all 157 accounted for.
- **A regression I introduced and caught on the Mac:** an MPS skip silently stopped firing.

**Comment by MakazhanAlpamys** [comment]
#1050 (merged as 2131b683), by @Srinivasan8888, lands the single CUDA probe and the `gpu` marker. I am leaving this issue open, because the PR says `Refs`, and the acceptance item it considers outstanding is a real `pytest -m gpu` card run recorded in `benchmarks/gpu-test-runs.md`.

Measured during review on an RTX 5070 (sm_120), for whoever records that run:
- `-m gpu --collect-only` selects 88 tests, all in the touched modules, and every one of them skips without CUDA.
- Whole-suite collection is identical with and without CUDA.
- Per-test JUnit comparison against `main` shows no test moving from passed to skipped.

**Comment by MakazhanAlpamys** [comment]
Last criterion closed. Ran it on the card: `2d1714ba`, row in `benchmarks/gpu-test-runs.md`.

```
pytest tests/ -m gpu --no-cov -p no:randomly -q
RTX 5070 Laptop (sm_120), 8123 MiB, driver 616.92 / CUDA 13.4,
torch 2.14.0+cu130, bitsandbytes 0.50.2, commit 4ae040fb

1 failed, 86 passed, 2 skipped, 23091 deselected in 106.11s
(reproduced: 98.90s, same single failure)
```

- [x] `pytest -m gpu --collect-only` selects every test that skips without CUDA — #1050
- [x] One CUDA probe in `conftest.py` plus the ratchet against per-module probes — #1050
- [x] `CONTRIBUTING.md` documents the command and the report format — #1050
- [x] `benchmarks/gpu-test-runs.md` holds a real `-m gpu` run, not only the two that predate the marker

**The one failure is order-dependent and pre-existing**, not a regression: `test_issue974_direct_range_reads.py::TestOnRealHardware::test_every_region_is_a_pinned_sector_aligned_view_of_a_power_of_two_arena` (`data_ptr % 4096`). It **passes alone** — 1 passed in 2.17s — and fails only after `test_issue971_async_disk_source.py` has run in the same process. Pinned-arena pressure, in my own streaming code, already recorded on bare `main`.

Both skips are correct: no `mlx` on this box, and `test_stream_multi_gpu_guard.py` needs a second real CUDA device.

**Three things worth recording, which only became visible once the marker existed:**

1. **The selection is 88 tests across 18 files**, and its shape is not what the issue title suggests. The biggest contributors are `test_v07204.py` (14), `test_issue971_async_disk_source.py` (11), `test_v07203.py` (10) and `test_issue968_bnb_nf4_path_divergence.py` (9) — layer streaming and the bitsandbytes 4-bit paths are most of what a card exercises in this repo.
2. **A green run here does NOT mean #776 is resolved, and the reason is worth knowing.** Those nine `test_issue968` tests pass by asserting that the two bitsandbytes 4-bit compute paths still **diverge**, by a non-zero amount within an order of magnitude of the recorded value. They are written to fail if the paths ever agree, which would be the signal to re-measure #776. So "86 passed" includes nine tests whose passing is evidence that an open defect is still there. I put that in the file, because it is exactly the kind of thing a future reader would get backwards.
3. **It runs in under two minutes** against roughly thirty for the full suite. That is the argument for asking every contributor with a card to run it: it is cheap, and CI structurally cannot — GitHub's runners have no GPU, so all 88 skip in all nine cells.

The estimate in this issue was "~51 CUDA-only tests"; the real number is 88. Closing.
